### DISCUSSION TITLE
Regeneration of tokens during the artillery test run.

### DISCUSSION THREAD

**Discussion by Mmallineni** [artilleryio/artillery#2326] [Artillery]
The access tokens I use exprire in 24 hrs, I want to run a 72 hours long test. How is it possible ? I can regenrate the tokens while the test is in run and whether the test be able to pick them up?

**Comment by BenderRules** [comment]
I have a similar issue. I don't want to run it for that long, but trying to execute a script through run-lambda, the setup takes so much time, that when the first call is done, I'm getting a "token expired" message. Our authentication remains for one hour through SSO login, and auto-renewal doesn't work : 

`"Error: Unable to refresh AWS credentials from SSO
    at updateSSOCredentials (C:\Users\%user%\AppData\Roaming\npm\node_modules\artillery\lib\platform\aws\aws-set-default-credentials.js:63:13)
    at async Timeout._onTimeout (C:\Users\%user%\AppData\Roaming\npm\node_modules\artillery\lib\platform\aws\aws-set-default-credentials.js:36:9)"`

**Reply by bernardobridge** [reply]
Hi @BenderRules 👋 ,

I'm not sure this is related to the originally posted issue. That being said, can you clarify how long it's taking when you say "the setup takes so much time"? The setup of the Lambda function to run a test should not take much longer than a couple of minutes.

Also, a couple of side notes:
- you can increase token expiration for roles in AWS to up to 12 hours.
- if you're planning on running longer tests we would recommend Fargate support instead (as Lambda has a 15 minute timeout).

**Reply by BenderRules** [reply]
Yes you are right. That is of course not the same issue. However, the preparation, such as the zip packaging and AWS environment preparation takes quite a while. Way more than a couple of minutes, so I don't know what is going on. I'm running it from my local machine, from Powershell. Unlike the OP, I don't require a long running test. Is there a debug mode, to make it a bit more verbose ?

**Comment by bernardobridge** [comment]
Hi @Mmallineni ,

Apologies, we missed this discussion at the time. When you say the access tokens you use, do you mean access tokens from the service you're trying to test? 

Artillery allows you to run custom javascript logic during the test, so you can always have something to refresh the token mid test if needed.

**Reply by sridharveeravalli** [reply]
I have used a similar approach and it works out for me for the most part. I further break it down using a loop construct to determine how frequently the token need to be refreshed using a custom JS function

**Comment by hassy** [comment]
We just added an example for generating + refreshing VU auth tokens as a test is running: https://github.com/artilleryio/artillery/tree/main/examples/refresh-auth-token
